hibernate-core/src/main/java/org/hibernate/query/Pager.java

```diff
@@ -0,0 +1,247 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or http://www.gnu.org/licenses/lgpl-2.1.html
+ */
+package org.hibernate.query;
+
+import org.hibernate.Incubating;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static java.util.Collections.emptyList;
+import static java.util.Spliterator.IMMUTABLE;
+import static java.util.Spliterator.NONNULL;
+import static java.util.Spliterator.ORDERED;
+import static java.util.Spliterators.spliteratorUnknownSize;
+
+/**
+ * Allows bidirectional iteration over pages of query results, useful for processing
+ * query results in batches.
```

Comment:
You might want to mention limitations. Depending on the isolation level, some results might be returned twice (in subsequent pages) or not at all... right?

I think comparing with `ScrollableResults` would make sense, too, as it wouldn't be immediately clear to a new user why there are two ways of doing something very similar.

Reply:
OK, yup, both good ideas. Thanks.